src/Service/URLSanitisationService.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+namespace SilverStripe\StaticPublishQueue\Service;
+
+use SilverStripe\Core\Config\Configurable;
+use SilverStripe\Core\Environment;
+
+/**
+ * Class URLSanitisationService
+ * Provides an abstraction for populating and generating a list of
+ * URLs to be cached by StaticPublishQueue.
+ *
+ * @package SilverStripe\StaticPublishQueue\Service
+ */
+class URLSanitisationService
+{
+    use Configurable;
+
+    /**
+     * Enable to force all URLs to be changed to https
+     *
+     * @var bool
+     * @config
+     */
+    private static $force_ssl = false;
+
+    /**
+     * @var array
+     */
+    private $urls = [];
+
+    /**
+     * @var int
+     */
+    private $pointer = 0;
+
+    /**
+     * @param string $url
+     */
+    public function addURL(string $url): void
+    {
+        // We never want to have stage=Stage present in our cached URLs. This will safely remove "stage" params, but
+        // keep any others. It doesn't matter where in the string "stage=" exists.
+        $url = preg_replace('/([?&])stage=[^&]+(&|$)/', '$1', $url);
```

Comment:
You never want to have `stage` as a part of URL for static publishing as we always statically publish `Live` content. When `stage` param is present it indicated the fact that static cache needs to be bypassed.

The need to remove `stage` param from URL became started when the session change was implemented in the framework. This change would make links add the `stage` param whenever a user is logged in. This is a problem for static publish module as it runs the jobs with a user as there was no other option available at that time.

In the meantime, I added an option to run a job without a user to the queued jobs module so this code may no longer be needed provided we run the job without user.

I'll look into this more and ideally this code gets removed.

Reply:
I couldn't find a way to ensure that `stage` param is not present because `urlsToCache()` is an interface so it's often up to the project code to deliver list of urls. The problem with `stage` param is caused by two things:

1 - running a job with user (this case is fixed in this PR)
2 - generating a link with `stage` param caused by Versioned (when default reading mode is different from current reading mode for some reason)